components/tikv_util/src/resizable_threadpool.rs

```diff
@@ -1,161 +1,350 @@
-    where
-        F: Future,
-    {
-        let inner = self.inner.read().unwrap();
-        inner.block_on(f)
+        self.count += 1;
+        let thread_name = format!("{}-v{}-{}", self.thread_name, self.count, new_size);
+
+        let new_runtime = (self.replace_pool_rule)(new_size, &thread_name)
+            .unwrap_or_else(|_| panic!("failed to create tokio runtime {}", thread_name));
+
+        let old_runtime = std::mem::replace(
+            &mut *runtime_guard,
+            DeamonRuntime {
+                inner: Some(new_runtime),
+                tracker: TaskTracker::new(),
+            },
+        );
+        self.gc_runtime.spawn(async move {
+            old_runtime.tracker.close();
+            old_runtime.tracker.wait().await;
+            drop(old_runtime);
+        });
+
+        info!(
+            "Resizing thread pool";
+            "thread_name" => thread_name.as_str(),
+            "new_size" => new_size
+        );
+
+        self.size = new_size;
+        (self.after_adjust)(new_size);
     }
 }
 
 #[cfg(test)]
 mod test {
-    use std::sync::atomic::{AtomicUsize, Ordering};
+    use std::{
+        sync::atomic::{AtomicUsize, Ordering},
+        thread::{self, sleep},
+        time::Duration,
+    };
 
     use super::*;
+    use crate::time::Instant;
 
-    static COUNTER: AtomicUsize = AtomicUsize::new(0);
+    fn replace_pool_rule(thread_count: usize, thread_name: &str) -> TokioResult<Runtime> {
+        let rt = tokio::runtime::Builder::new_multi_thread()
+            .worker_threads(thread_count)
+            .thread_name(thread_name)
+            .enable_all()
+            .build()
+            .unwrap();
+        Ok(rt)
+    }
 
     #[test]
     fn test_adjust_thread_num() {
-        let replace_pool_rule = |thread_count: usize, thread_name: &str| {
-            let rt = tokio::runtime::Builder::new_multi_thread()
-                .worker_threads(thread_count)
-                .thread_name(thread_name)
-                .enable_all()
-                .build()
-                .unwrap();
-            Ok(rt)
+        static COUNTER: AtomicUsize = AtomicUsize::new(4);
+        let after_adjust = |new_size: usize| {
+            COUNTER.store(new_size, Ordering::SeqCst);
         };
+        let mut threads = ResizableRuntime::new(
+            COUNTER.load(Ordering::SeqCst),
+            "test",
+            Box::new(replace_pool_rule),
+            Box::new(after_adjust),
+        );
+        assert_eq!(COUNTER.load(Ordering::SeqCst), threads.size());
+
+        let handle = threads.handle();
+        handle.block_on(async {
+            COUNTER.fetch_add(1, Ordering::SeqCst);
+        });
+        assert_eq!(COUNTER.load(Ordering::SeqCst), threads.size() + 1);
+
+        threads.adjust_with(8);
+        assert!(!threads.gc_runtime.tracker.is_empty());
+        assert_eq!(COUNTER.load(Ordering::SeqCst), threads.size());
+
+        sleep(Duration::from_secs(11));
```

Comment:
err you will make the EE team very unhappy with this. can the test run without sleeping for 11 seconds?